altair/vegalite/v5/api.py

```diff
@@ -345,11 +384,596 @@ def check_fields_and_encodings(parameter: Parameter, field_name: str) -> bool:
+        return _ChainedWhen(condition, self.to_dict())
+
+    def to_dict(self, *args: Any, **kwargs: Any) -> _Conditions:
+        return self._conditions.copy()
+
+
+class _ChainedWhen(_BaseWhen):
+    """
+    Utility class for `when-then-otherwise` conditions.
+
+    Represents the state after calling `alt.when(...).then(...).when(...)`.
+
+    This partial state requires calling `then` to finish the condition.
+
+    References
+    ----------
+    [polars.expr.whenthen](https://github.com/pola-rs/polars/blob/b85c5e0502ca99c77742ee25ba177e6cd11cf100/py-polars/polars/expr/whenthen.py)
+    """
+
+    def __init__(self, condition: _ConditionType, conditions: _Conditions, /) -> None:
+        self._condition = condition
+        self._conditions = conditions
+
+    @overload
+    def then(
+        self,
+        statement: _StatementOrLiteralType,
+        *,
+        str_as_lit: bool = ...,
+        seq_as_lit: Literal[True] = ...,
+        **kwargs: Any,
+    ) -> _Then: ...
+    @overload
+    def then(
+        self,
+        statement: _StatementType,
+        *,
+        str_as_lit: bool,
+        seq_as_lit: Literal[False],
+        **kwargs: Any,
+    ) -> _Then: ...
+    def then(
+        self,
+        statement: _StatementOrLiteralType,
+        *,
+        str_as_lit: bool = False,
+        seq_as_lit: bool = True,
+        **kwargs: Any,
+    ) -> _Then:
+        lit = _LiteralConfig(str_as_lit, seq_as_lit)
+        condition = self._when_then(statement, kwargs, lit)
+        conditions = self._conditions.copy()
+        conditions["condition"].append(condition)
+        return _Then(conditions)
+
+
+def _when(
+    predicate: _AltOptional[_PredicateType] = Undefined,
+    *more_predicates: _ComposablePredicateType,
+    empty: _AltOptional[bool] = Undefined,
+    **constraints: Any,
+) -> _When:
+    """Start a `when-then-otherwise` condition.
+
+    Notes
+    -----
+    Directly inspired by the `when-then-otherwise` syntax used in `polars.when`.
+
+    Similar to `alt.condition` but:
+
+    Allows specifiying a predicate, without a default:
+    ```py
+    condition = (predicate, if_true)
+    ```
+
+    Allows multiple precise predicates, optionally ending with a default:
+    ```py
+    conditions = [
+        (predicate_1, if_true_1),
+        (predicate_2, if_true_2),
+        (predicate_3, if_true_3),
+        ...,
+        if_false,
+    ]
+    ```
+
+    References
+    ----------
+    [polars.when](https://docs.pola.rs/py-polars/html/reference/expressions/api/polars.when.html)
+    """
+    condition = _parse_when(predicate, *more_predicates, empty=empty, **constraints)
+    return _When(condition)
+
+
+def _condition_to_expr_str(c: Union[_SelectionType, _Conditions], /) -> str:
+    """Convert the output of `alt.condition`/`when-then-otherwise` to an expression string."""
+    return _condition_to_expr_ref(c).expr
+
+
+def _condition_to_expr_ref(c: Union[_SelectionType, _Conditions], /) -> core.ExprRef:
```

Comment:
@mattijn [comment](https://github.com/vega/altair/pull/3427#issuecomment-2144027795)
> Hmm, maybe we should add  `alt.expr.when()` for this?

Although this wasn't part of the original issue, this was where I landed on this idea.

From my understanding, `alt.expr.if_` and vega-lite [if](https://vega.github.io/vega/docs/expressions/#if) are more limited in that they always require 3 arguments. 
So while `when-then-otherwise` seems a good fit for this case, `alt.expr.if_` couldn't accept:
- `when-then`
- `when-then-when-then-otherwise`
- etc

Personally, I'd expect the same behaviour from both and find it frustrating to encounter the limitation.

An option could be accepting a subset of possible outputs from `alt.condition`/`when-then-otherwise`. At least for the two examples you provided, this solution could work. 

I must admit though I'm not familiar with `alt.expr` in general, and found it difficult to follow outside of a notebook environment.

Reply:
I’m fine with leaving any work done on `alt.expr.when()` out of this PR @dangotbanned! We can open a new issue on this topic, based on the experience of this PR. I’m positive we can figure out a solution, since the work done in this PR is more advanced than what is needed for `alt.expr.if_`.

Reply:
Removed in [`f89990f` (#3427)](https://github.com/vega/altair/pull/3427/commits/f89990f512e12bd11d3f511900410840f139f081)